# Scrape apartments.com 

let's look at East Hollywood for apartments.com. 

## Search page 

In [44]:
import pandas as pd
import geopandas as gpd

In [45]:
import requests
from bs4 import BeautifulSoup

In [46]:
url_eho = 'https://www.apartments.com/east-hollywood-los-angeles-ca/'

In [47]:
import time
import random

In [6]:
#to look at the structure of apartments.com 

time.sleep(random.random()*5)

r = requests.get(url_eho, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:68.0) Gecko/20100101 Firefox/68.0'})

From Adam: 

The website is probably blocking you. It might be detecting that you are calling the webpage from Python.

The solution is to set a UserAgent. Google around to see more. But I would be careful: in particular, limit the pace at which you scrape. For example, this will add a random gap of between 0 and 5 seconds:

SOURCE: https://www.kite.com/python/answers/how-to-make-a-request-with-a-user-agent-in-python#:~:text=Use%20requests.,request%20from%20the%20source%20url%20. 
headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"}


In [11]:
soup_eho = BeautifulSoup(r.content)
print(soup_eho.prettify())

<!DOCTYPE html>
<html class="fontBaseSizeV2" data-placeholder-focus="false" id="appContainer" lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport"/>
  <meta content="telephone=no" name="format-detection"/>
  <meta content="email=no" name="format-detection"/>
  <meta content="true" name="HandheldFriendly"/>
  <title>
   East Hollywood Apartments for Rent - Los Angeles, CA | Apartments.com
  </title>
  <link href="/a/8a55f9/favicon.ico" rel="icon"/>
  <link href="https://www.apartments.com/east-hollywood-los-angeles-ca/" rel="canonical">
   <meta content="See all 1,731 apartments in East Hollywood, Los Angeles, CA currently available for rent. Check rates, compare amenities and find your next rental on Apartments.com." name="description"/>
   <meta content="en" name="language"/>
   <meta content="en" http-equiv="language"/>
   <meta content="en-US" http-equiv="content-language"/>
   <meta

## Listing page
For one particular East Hollywood listing: 

In [7]:
url_listing = 'https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/'

In [8]:
time.sleep(random.random()*7)
r_listing = requests.get(url_listing, headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"})

In [9]:
soup_listing = BeautifulSoup(r_listing.content)
print(soup_listing.prettify())

<!DOCTYPE html>
<html class="fontBaseSizeV2" data-placeholder-focus="false" id="appContainer" lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" name="viewport"/>
  <meta content="telephone=no" name="format-detection"/>
  <meta content="email=no" name="format-detection"/>
  <meta content="true" name="HandheldFriendly"/>
  <title>
   Lido Apartments - Los Angeles, CA | Apartments.com
  </title>
  <link href="/a/8a55f9/favicon.ico" rel="icon"/>
  <link href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/" rel="canonical">
   <meta content="See all available apartments for rent at Lido Apartments in Los Angeles, CA. Lido Apartments has rental units ranging from 660-1116 sq ft starting at $2195. " name="description"/>
   <meta content="en" name="language"/>
   <meta content="en" http-equiv="language"/>
   <meta content="en-US" http-equiv="content-language"/>
   <meta content="73929927

 Each listing is in the following text: 
 
 <li class="mortar-wrapper">
          <article class="placard placard-option has-header js-diamond" data-ck="txmsmyc" data-listingid="9nkznr5" data-url="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/">
           <header class="placard-header has-logo">
            <div class="property-information">
             <a class="property-link" href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/">
              <div class="property-title" title="Lido Apartments, Los Angeles, CA">
               <span class="js-placardTitle title">
                Lido Apartments
               </span>
              </div>
              <div class="property-address js-url" title="4847 Oakwood Ave, Los Angeles, CA 90004">
               4847 Oakwood Ave, Los Angeles, CA 90004
              </div>
             </a>
            </div>
            <div class="property-logo-wrapper">
             <a class="property-link js-url" href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/">
              <div class="property-logo-inner">
               <div aria-label="Lido Equities Management Group, Inc." class="property-logo" role="img" style="background-image: url(https://images1.apartments.com/i2/qziBm5P3aq38bOLTndPUd_L_fZEwKKv1gP9hs-RP8eQ/110/lido-equities-management-group-inc-logo.png);">
               </div>
              </div>
             </a>
            </div>
            <div class="favorite-wrapper">
             <a class="favoriteIcon neutral" data-liked-classes="liked" data-neutral-classes="neutral" href="javascript:void(0);">
             </a>
            </div>
           </header>
           <section class="placard-content">
            <div class="content-inner">
             <div class="media-wrapper">
              <div class="media-sizer">
              </div>
              <div class="media-outer">
               <div class="media-inner">
                <div class="media">
                 <div class="placardCarouselImgCount">
                  <span class="js-spnImgNum">
                   1
                  </span>
                  /
                  <span class="js-spnImgCount">
                   14
                  </span>
                 </div>
                 <div class="galleryControl">
                  <button class="imageCarouselLeftContainer" type="button">
                   <i class="imageCarouselArrowLeftIcon">
                   </i>
                  </button>
                  <button class="imageCarouselRightContainer" type="button">
                   <i class="imageCarouselArrowRightIcon">
                   </i>
                  </button>
                 </div>
                 <div class="imageContainer carousel slide" data-interval="false">
                  <div class="carouselSpinner">
                   <i class="aptsLogoMarkIcon rotate">
                   </i>
                  </div>
                  <div class="fallbackSpinner">
                   <i class="spinner">
                   </i>
                  </div>
                  <div class="carouselInner">
                   <div class="item active" style='background-image: url("https://images1.apartments.com/i2/ZyaYVNy1sPwc1FjhLNVkkeZmPK8kBDlxD5q-a4vWefc/117/lido-apartments-los-angeles-ca-building-photo.jpg");' title="Building Photo - Lido Apartments">
                   </div>
                  </div>
                 </div>
                 <div class="placardBanner">
                  <button class="videoPlay" data-datasource="42" data-video="https://vapi.costargroup.com/video/play/0u6i3f?source=42" data-videokey="0u6i3f" title="Watch Lido Apartments Video" type="button">
                   <i class="playVideoPlacardIcon">
                   </i>
                  </button>
                  <button class="virtualTour" title="See Lido Apartments in 3D" type="button">
                   <span>
                    3D Tours
                   </span>
                  </button>
                 </div>
                </div>
               </div>
               <div class="mediaLinks">
                <ul class="mediaLinksList">
                 <li>
                  <a href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/#photos">
                   14 Photos
                  </a>
                 </li>
                 <li>
                  <a href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/#3dtours">
                   Virtual Tour
                  </a>
                 </li>
                 <li>
                  <a href="https://www.apartments.com/lido-apartments-los-angeles-ca/9nkznr5/#videos">
                   Videos
                  </a>
                 </li>
                </ul>
               </div>
              </div>
             </div>
             <div class="property-info">
              <div class="content-wrapper">
               <div class="property-details-wrapper has-header has-specials">
                <span class="listingFreshness">
                 <span class="lastUpdated">
                  <i class="rotateRightIcon" title="Last Updated">
                  </i>
                  <span>
                   2 days Ago
                  </span>
                 </span>
                </span>
                <div class="specials" data-specials-label="Specials">
                 Specials
                 <i class="storyicon saleStoryIcon">
                 </i>
                 <span class="specialOfferTooltip">
                  This property has a special promotion.
                 </span>
                </div>
               </div>
               <div class="price-wrapper">
                <div class="price-range">
                 $2,195 - 3,570
                </div>
               </div>
               <div class="property-information-wrapper">
                <div class="bed-range">
                 1-3 Bed
                </div>
                <div class="availability">
                 Avail. Soon
                </div>
               </div>
               <div class="property-amenities">
                <span>
                 Dog Friendly
                </span>
                <span>
                 Cat Friendly
                </span>
                <span>
                 Air Conditioning
                </span>
                <span>
                 Washer/Dryer - In Unit
                </span>
                <span>
                 Dishwasher
                </span>
               </div>
               <div class="phone-wrapper">
                <a class="phone-link js-phone" href="tel:6282306839">
                 <i class="storyicon comMobileStoryIcon">
                 </i>
                 <span>
                  628-230-6839
                 </span>
                </a>
               </div>
               <div class="property-actions">
                <div class="actions-wrapper">
                 <button class="checkAvailability btn">
                  Email
                 </button>
                </div>
               </div>
              </div>
             </div>
            </div>
           </section>
          </article>
         </li>

## Function 

In [48]:
def getPosts(neighborhood, howmany):
    postList = []
    n = neighborhood.replace(' ', '-')
    
    # range 0 (1st page) - number of posts you want, going up in 120 increments 
    ## looping over the pages until you reach the target HOWMANY posts 
    for i in range(1, howmany, 25): 
        print(i)
        
    #FOR EACH PAGE: 
        url = f'https://www.apartments.com/{n}-los-angeles-ca/{i}/'  
        #request the text from the url 
        time.sleep(random.random()*5)
        r =requests.get(url, headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"})
        soup = BeautifulSoup(r.content)

        #get a list of posts 
        posts = soup.find_all('li', class_="mortar-wrapper")
        
        #FOR EACH POST 
        for post in posts: 
            #name
            name = post.find('span', class_="js-placardTitle title")
            if name is not None: 
                name = name.text
 #           print(name)
            
            #address
            addy = post.find('div', class_="property-address js-url")
            if addy is not None: 
                addy = addy.text
#            print(addy)
            
            #property management
            mgmt = post.find('div', class_="property-logo-inner")
            if mgmt is not None:
                mgmt = mgmt.text
#            print(mgmt)
            
            #phone number 
            phone = post.find('a', class_="phone-link js-phone")['href']  
            if phone is not None:
                phone = phone 
#            print(type(phone))

            #price 
            price = post.find('div', class_= "price-range")
            if price is not None: 
                price = price.text 
#            print(type(price))
            
            #unit type - each post has a table of unit type and price, do that in another function
            beds = post.find('div', class_="bed-range")   
            if beds is not None:
                beds = beds.text 
#            print(type(beds))

            #amenities
            amenities = post.find('li', class_="specInfo uniqueAmenity")
            if amenities is not None:
                amenities = amenities.text 
#            print(type(amenities)) 


###START HERE
            #go into the individual post's url and get the description...
            url_post = post.find('a', class_="property-link")['href']
            time.sleep(random.random()*7)
            r_post = requests.get(url_post, headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"})
            soup_post = BeautifulSoup(r_post.content)
            description = soup_post.find('section', id="descriptionSection")  

#             latlon = soup_post.find('div', class_='viewposting')
#             if latlon is not None:
#                 latitude = latlon['data-latitude']
#                 longitude = latlon['data-longitude']
            
            postList.append({'Listing':name, 'Address':addy, 'Description':description,'Price': price, 'Unit Type':beds, 'Amenities':amenities, 'Property Manager': mgmt, 'Phone': phone, 'URL':url_post})
            # missing: 'lat': latitude, 'lon': longitude
            
        df = pd.DataFrame(postList)
        
    return df

In [49]:
eho_25 = getPosts('East Hollywood', 25)

1


In [16]:
eho_500 = getPosts('East Hollywood', 500)
### 500 --> NoneType is not scriptable 
### do a try except:? 

1


In [19]:
eho_25.head()

Listing                                   Address  \
0          Lido Apartments   4847 Oakwood Ave, Los Angeles, CA 90004   
1         AVA Toluca Hills   3600 Barham Blvd, Los Angeles, CA 90068   
2         Violet on Virgil   160 S Virgil Ave, Los Angeles, CA 90004   
3          eaves Los Feliz  3100 Riverside Dr, Los Angeles, CA 90027   
4  ST Townhomes at Romaine    5066 Romaine St, Los Angeles, CA 90029   

                                         Description           Price  \
0  [\n, [About Lido Apartments], \n, [A Lido Apar...  $2,195 - 3,570   
1  [\n, [About AVA Toluca Hills], \n, [AVA Toluca...  $1,665 - 3,210   
2  [\n, [About Violet on Virgil], \n, [Violet on ...  $1,500 - 2,450   
3  [\n, [About eaves Los Feliz], \n, [The best of...  $1,785 - 2,912   
4  [\n, [About ST Townhomes at Romaine], \n, [Are...  $2,385 - 2,600   

        Unit Type                                          Amenities  \
0         1-3 Bed  \nDog Friendly\nCat Friendly\nAir Conditioning...   
1  Studio - 2 Bed  \nDog Friendly\nCat Friendly\nAir Conditioning...   
2  Studio - 2 Bed  \nDog Friendly\nCat Friendly\nAir Conditioning...   
3  Studio - 2 Bed  \nDog Friendly\nCat Friendly\nAir Conditioning...   
4           1 Bed  \nDog Friendly\nAir Conditioning\nWasher/Dryer...   

  Property Manager           Phone  \
0             \n\n  tel:6282306839   
1             \n\n  tel:8443653378   
2             \n\n  tel:8664606358   
3             \n\n  tel:8442562676   
4             \n\n  tel:8332494045   

                                                 URL  
0  https://www.apartments.com/lido-apartments-los...  
1  https://www.apartments.com/ava-toluca-hills-lo...  
2  https://www.apartments.com/violet-on-virgil-lo...  
3  https://www.apartments.com/eaves-los-feliz-los...  
4  https://www.apartments.com/st-townhomes-at-rom...

In [20]:
eho_25["Description"].loc[0]

<section class="descriptionSection js-viewAnalyticsSection" data-analytics-name="about" id="descriptionSection">
<h2 class="sectionTitle">About Lido Apartments</h2>
<p>A Lido Apartments Community

Brand New Modern Apartment Homes! 
Large Two &amp; Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more! 

 Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. 

Gated Parking and Gated Entry. A

In [6]:
#The description has the Unique Amenities listed in print(<li class = "specInfo uniqueAmenity">)! 

# NLP 

In [50]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize, sent_tokenize
nltk.download('stopwords')
nltk.download('punkt')


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\audre\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\audre\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

### 1. Create a long string of the Descriptions of each listing scraped 
Use regex library 

In [8]:
# not sure why this isn't working... 
descriptions_str = ""
for listing in eho_25:
    descriptions_str += str(listing.Description)
print('Text is {} characters long'.format(len(descriptions_str)))

AttributeError: 'str' object has no attribute 'Description'

In [51]:
#Define new variable that is the description column turned into strings --> a series of str
bigstr_25 = eho_25.Description.apply(str)
bigstr_25

0     <section class="descriptionSection js-viewAnal...
1     <section class="descriptionSection js-viewAnal...
2     <section class="descriptionSection js-viewAnal...
3     <section class="descriptionSection js-viewAnal...
4     <section class="descriptionSection js-viewAnal...
5     <section class="descriptionSection js-viewAnal...
6     <section class="descriptionSection js-viewAnal...
7     <section class="descriptionSection js-viewAnal...
8                                                  None
9     <section class="descriptionSection js-viewAnal...
10    <section class="descriptionSection js-viewAnal...
11    <section class="descriptionSection js-viewAnal...
12    <section class="descriptionSection js-viewAnal...
13    <section class="descriptionSection js-viewAnal...
14    <section class="descriptionSection js-viewAnal...
15    <section class="descriptionSection js-viewAnal...
16    <section class="descriptionSection js-viewAnal...
17    <section class="descriptionSection js-view

series

In [52]:
# turn the series of str into a list 
bigstr_25 = bigstr_25.tolist()
bigstr_25

['<section class="descriptionSection js-viewAnalyticsSection" data-analytics-name="about" id="descriptionSection">\n<h2 class="sectionTitle">About Lido Apartments</h2>\n<p>A Lido Apartments Community\n\nBrand New Modern Apartment Homes! \nLarge Two &amp; Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more! \n\n Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. \n\nGated Parking and Gat

list

In [55]:
bigstr_25 = str(bigstr_25)
bigstr_25

'[\'<section class="descriptionSection js-viewAnalyticsSection" data-analytics-name="about" id="descriptionSection">\\n<h2 class="sectionTitle">About Lido Apartments</h2>\\n<p>A Lido Apartments Community\\n\\nBrand New Modern Apartment Homes! \\nLarge Two &amp; Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more! \\n\\n Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. \\n\\nGated Park

string! this adds an extra \ 

### 2. Remove Tags 

Method 1. re.sub('<[^>]*>', '', mystring)

Remove the tags using code from SOURCE: https://stackoverflow.com/questions/3662142/how-to-remove-tags-from-a-string-in-python-using-regular-expressions-not-in-ht


In [56]:
bigstr_25_w = re.sub('<[^>]*>', '', bigstr_25_)
bigstr_25_w

'[\'\\nAbout Lido Apartments\\nA Lido Apartments Community\\n\\nBrand New Modern Apartment Homes! \\nLarge Two &amp; Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more! \\n\\n Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. \\n\\nGated Parking and Gated Entry. ALL 3 bedroom units come with 3 parking spaces! We don’t charge extra for parking!\\n\\nGrand Opening August 1, 2020!! \\n\\

Method 2. beautiful soup - does the same thing

In [58]:
# remove the tags...
str_soup = BeautifulSoup(str(bigstr_25))
bigstr_25_w = str_soup.get_text() 

### 3. Remove \n

Now get rid of the new lines

In [61]:
bigstr_25_w = re.sub(r"\\n", "", bigstr_25_w)
bigstr_25_w

'[\'About Lido ApartmentsA Lido Apartments CommunityBrand New Modern Apartment Homes! Large Two & Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more!  Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. Gated Parking and Gated Entry. ALL 3 bedroom units come with 3 parking spaces! We don’t charge extra for parking!Grand Opening August 1, 2020!! Pre Leasing available now.Unique FeaturesB

Now get rid of the white space! 

In [62]:
bigstr_25_w = re.sub(r"\s+", " ", bigstr_25_w)
bigstr_25_w 


'[\'About Lido ApartmentsA Lido Apartments CommunityBrand New Modern Apartment Homes! Large Two & Three Bedrooms all with stylish, top quality finishes, designer kitchens with custom cabinetry w/ self closing doors and drawers, high end appliances including refrigerators, dishwashers, gas stoves, electric ovens, microwaves, dishwashers and washer/dryers. oversized sinks, wood vinyl floors throughout, stylish bathrooms with Italian tiles, oversized bathtubs balconies and more! Building amenities include Fitness Center, Game Room with bar, large flatscreen televisions, ping pong table, foosball table, business center, rooftop deck with incredible City Mountain Views, outdoor television, built in barbecue bars, fire-pits, sun tanning decks with lounge chairs and tables... a really cool place to hang out. Gated Parking and Gated Entry. ALL 3 bedroom units come with 3 parking spaces! We don’t charge extra for parking!Grand Opening August 1, 2020!! Pre Leasing available now.Unique FeaturesBa

Now we're ready to find the word frequencies. 

### 2. Wordcounter function to find the frequency of words in this list. 

In [63]:
import pandas as pd 

def wordcounter(wordlist):
    counts = {} #dictionary to hold counts of each word
    for word in wordlist:
        lword = word.lower()
        if lword in counts: 
            counts[lword] += 1 
        else:
            counts[lword] = 2
    #dictionary --> dataframe        
    df = pd.DataFrame.from_dict(counts, orient = 'index', columns=['word_count']) #defines a new word_count column for the df
    df.sort_values('word_count', ascending=False, inplace=True)
    df.index.name = 'word'
    return df 

In [64]:
#create a list of all the words, splitting on the spaces. 
wordcounter(bigstr_25_w).head(50)

word_count
word            
            4566
e           2980
o           2157
t           2146
a           2049
i           2020
n           2003
s           1717
r           1703
l           1392
d            970
c            931
h            876
u            831
m            685
f            592
p            574
g            535
y            529
w            463
b            412
,            382
v            346
k            246
-            202
.            199
x             74
*             68
0             67
q             66
1             60
'             57
2             57
/             55
z             54
!             43
’             37
4             35
:             35
j             31
5             26
)             25
(             25
$             20
3             20
9             20
8             19
&             18
6             18
7             11

None


In [32]:
# sara's code - Initial_Scraping notebook 
# another way of doing this
# solution
# method 1: modify the function
def countWords_nostop(wordlist):
    
    # this might be overkill, but we probably want a version of stopwords that excludes the punctuation  
    # this is a list comprehension - a shorthand for a loop
    # original code: swords = [re.sub(r"[^A-z\s]", "", sword) for sword in stopwords.words('english')]
    swords = stopwords.words('english')
    swords.append('br')
    swords.append('div')
    swords.append('la')
    swords.append('los')
    swords.append('angeles')
    
    # in longhand, that was
    # swords = []
    # for sword in stopwords.words('english'):
    #     swords.append(re.sub(r"[^A-z\s]", "", sword))
    
    counts = {} # a dictionary to hold the counts
    for word in wordlist:
        lword = word.lower()
        if lword in swords:
            # skip the stop words
            continue
        elif lword in counts:
            counts[lword] +=1
        else:
            # doesn't exist in the dictionary
            counts[lword] = 1

    # convert the dictionary to a dataframe
    # https://stackoverflow.com/questions/18837262/convert-python-dict-into-a-dataframe
    df = pd.DataFrame.from_dict(counts, orient='index', columns=['word_count'])

    # sort it by the word_count column
    df.sort_values('word_count', ascending=False, inplace=True)
    df.index.name = 'word'

    return df

df = countWords_nostop(wordlist)
print(df.head())

TypeError: 'method' object is not iterable

In [30]:
# now lets create a function that does a word count for the words in the wordsonly df
def countWords(wordlist):
    counts = {} # a dictionary to hold the counts
    for word in wordlist:
        lword = word.lower() # convert to lowercase
        if lword in counts:
            counts[lword] +=1
        else:
            # doesn't exist in the dictionary
            counts[lword] = 1

    # convert the dictionary to a dataframe
    # https://stackoverflow.com/questions/18837262/convert-python-dict-into-a-dataframe
    df = pd.DataFrame.from_dict(counts, orient='index', columns=['word_count'])

    # sort it by the word_count column
    df.sort_values('word_count', ascending=False, inplace=True)
    df.index.name = 'word'
    return df

# create a list of all the words, through splitting on the spaces
wordlist = mylines_words.split()
df = countWords(wordlist)

NameError: name 'mylines_words' is not defined

In [ ]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize, sent_tokenize
nltk.download('stopwords')
nltk.download('punkt')

In [ ]:
# create a pandas datafram that counts the words and lets us know if its a stop word or not
swords = [re.sub(r"[^A-z\s]", "", sword) for sword in stopwords.words('english')]

df = countWords(wordlist)
df.reset_index(inplace=True) # we want to do an apply on word, so it has to be a regular column, not the index

df['stopword'] = df.word.apply(lambda x: True if x in swords else False)
print(df.head(10))

df = df[df.stopword==False] # filter the dataframe to only be non-stopword words
print(df.head(10))

In [21]:
#define function that drops the tags and apply to the  
test_bald = re.sub('<section class="descriptionSection js-viewAnalyticsSection" data-analytics-name="about" id="descriptionSection"> <h2 class="sectionTitle">', '', test)
test_bald

TypeError: expected string or bytes-like object

In [8]:
apt_eastholly["Amenities"].loc[0]

'\nDog Friendly\nCat Friendly\nAir Conditioning\nWasher/Dryer - In Unit\nDishwasher\n'